keep/rulesengine/rulesengine.py

```diff
@@ -0,0 +1,127 @@
+                        f"Rule {rule.name} on event {event.id} is not relevant"
+                    )
+
+        self.logger.info("Rules ran, creating alerts")
+        # Create the alerts if needed
+        for rule in relevent_rules_for_events:
+            self.logger.info(f"Running relevant rule {rule.name}")
+            rule_results = self._run_rule(rule)
+            if rule_results:
+                self.logger.info("Rule applies, creating grouped alert")
+                # create grouped alert
+                event_payload = []
+                fingerprints = []
+                for group in rule_results:
+                    # todo: we take here the first but we should take all of them down the road
+                    group_payload = rule_results[group][0].dict().get("event")
+                    event_payload.append(group_payload)
+                    fingerprints.append(rule_results[group][0].fingerprint)
+                # TODO: should be calculated somehow else
+                fingerprint = hashlib.sha256(
+                    "".join(fingerprints).encode("utf-8")
+                ).hexdigest()
+                group_alert_name = f"Group alert {rule.name}: " + ", ".join(
+                    [event["name"] for event in event_payload]
+                )
+                create_alert_db(
+                    tenant_id=self.tenant_id,
+                    provider_type="rules",
+                    provider_id=rule.id,
+                    # todo: event should support list?
+                    event={
+                        "events": event_payload,
+                        "name": group_alert_name,
+                        "lastReceived": max(
+                            [event["lastReceived"] for event in event_payload]
+                        ),
+                        "severity": max([event["severity"] for event in event_payload]),
+                        "source": list(
+                            set([event["source"][0] for event in event_payload])
+                        ),
+                        # TODO: should be calculated somehow else
+                        "id": fingerprint,
+                        "status": "firing",
+                    },
+                    fingerprint=fingerprint,
+                )
+                self.logger.info("Created alert")
+
+        self.logger.info("Rules ran, alerts created")
+        return
+
+    def _extract_subrules(self, expression):
+        # CEL rules looks like '(source == "sentry") && (source == "grafana" && severity == "critical")'
+        # and we need to extract the subrules
+        sub_rules = expression.split(") && (")
+        # the first and the last rules will have a ( or ) at the beginning or the end
+        sub_rules[0] = sub_rules[0][1:]
+        sub_rules[-1] = sub_rules[-1][:-1]
+        return sub_rules
+
+    # TODO: a lot of unit tests to write here
+    def _run_rule_on_event(self, rule, event: AlertDto):
```

Comment:
i'm a bit confused about `_run_rule_on_event` and `_run_rule`, if I'm not mistaken, this "validates" whether the event is **relevant** for a rule? and `_run_rule` actually runs the rule?